feat(BusStopChange): store alerts into an S3 bucket for posterity

thecristen opened the pull request:
#### Summary of changes

<!-- Link to relevant Asana task; remove if not applicable -->
**Asana Ticket:** [Implement Bus Stop Change Draft w/ Historical Data](https://app.asana.com/0/385363666817452/1203783926700994/f)


Despite working on this application for years, this is my first time really tackling anything [OTP](https://serokell.io/blog/elixir-otp-guide)-related.

# Goal

We want to be able to display alerts relating to bus stop moves or closures from the past four ratings on the `/bus-stop-changes` page . We don't have great data readily available* for past alerts, and it seems easier to start saving and managing this data ourselves.

*(there's past alert information _out there_ in various places, but if we want to use our typical functions for displaying alerts we'd have to convert the format, and if we wanted to group them by routes or stops we're SOL if those are absent from the V3 API)

# Approach

It turns out we already have a `GenServer` that's fetching and parsing all the V3 API alerts every minute in `Alerts.Cache.Fetcher`. And since it's already structured using dependency injection, it was easy enough to swap out its arguments to construct an appropriate fetcher for this goal's use case:

| Argument | Default value | This PR uses |
| --- | --- | --- |
| Module/function/argument | `V3Api.Alerts.all()` | `V3Api.Alerts.all(args)` |
| Repeat (milliseconds) | `60_000` | `3_600_000` |
| Function to run on parsed alerts | `Alerts.Cache.Store.update/2` | `Alerts.Cache.BusStopChangeS3.copy_alerts_to_s3/2` |

That last function will handle saving the alert by doing the following:

* Filters the alerts to we actually care about (having an `effect` of `:stop_closure` or `:stop_moved`)
* Converting each alert to a new structure, `%HistoricalAlert{}`, that includes some helpful metadata for later display (the routes, stops, and municipality the alert affects)
* Saving to the S3 bucket, using the alert ID as the name. This gets rewritten on subsequent writes, ensuring that what ends up in S3 is always the latest revision of an alert.

There's also a function, `Alerts.Cache.BusStopChangeS3.get_stored_alerts/0` for retrieving _all_ saved alerts from S3.

I've also had to modify the template/view/controller for the `/bus-stop-changes` page to be able to handle displaying `%HistoricalAlert{}`.

---

#### General checks
* [x] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [x] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [ ] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

#### New endpoints, or non-trivial changes to current endpoints
~~* [ ] **Have we load-tested** any new pages or internal API endpoints that will receive significant traffic? [See load testing docs](./../apps/site/load_tests/README.md)~~
~~* [ ] **If this change involves routes,** does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?~~

paulswartz reviewed (commented):
Depending on what kind of filtering you need to do, we might consider something like DynamoDB instead of S3. It's slightly more complex to set up, but allows more fine grained queries.

thecristen commented on apps/site/lib/site_web/controllers/bus_stop_change_controller.ex:
```diff
@@ -38,16 +38,34 @@ defmodule SiteWeb.BusStopChangeController do
+    # Best guess at which alerts to designate as "past"
+    |> Enum.filter(fn %HistoricalAlert{alert: %Alert{active_period: [{_starting, ending} | _]}} ->
+      has_ended? = ending && Timex.before?(ending, yesterday)
+      is_nil(ending) || has_ended?
```
It's a best guess — there are lots of "indefinite" alerts around with no completion date noted before it dropped out of the V3 API current alerts results. This is especially likely when it involves routes/stops that end up deleted from our GTFS. 

Here are a few examples of such alerts (this screenshot shows alerts were from the Alerts UI export, and they aren't retrievable via V3 API)

<img width="884" alt="image" src="https://user-images.githubusercontent.com/2136286/228617150-3f101e4d-9562-415d-9f84-9d5a1a71e489.png">

thecristen commented on apps/alerts/lib/alert.ex:
```diff
@@ -50,8 +50,9 @@ defmodule Alerts.Alert do
 
   @type severity :: 0..10
   @type lifecycle :: :ongoing | :upcoming | :ongoing_upcoming | :new | :unknown
+  @type id_t :: String.t()
```
It's a pattern we have in other spots across the codebase, `Routes.Route.id_t()`, `Stops.Stop.id_t()`, `Schedules.Trip.id_t()`, etc, so I'm inclined to leave it in unless there are any disadvantages to using it!

thecristen commented on apps/alerts/lib/cache/bus_stop_change_s3.ex:
```diff
@@ -0,0 +1,119 @@
+    with stop_change_alerts when stop_change_alerts != [] <-
+           Enum.filter(bus_alerts, &(&1.effect in [:stop_closure, :stop_moved])) do
+      Enum.map(stop_change_alerts, & &1.id)
+      |> Enum.sort()
```
Maybe not - would the cache key match if the same alerts were sorted differently (I assumed no)? How confident should I be that the previous functions will maintain alerts in a consistent sort order (I assumed not very, but I might be wrong there)?

paulswartz replied:
Yes, you're right that the cache key wouldn't match. However, this also means that you won't re-write the alerts to S3 if the contents of the alert changes.

thecristen replied:
Yep!! I'm using `ttl: :timer.hours(24)` with the understanding that the cached value will expire then, and this will end up re-writing the alerts the next day. 

This reflects that I don't need to capture _very_ timely alert updates, but does expose me to the risk that an alert will make a meaningful change < 24h before concluding or being removed from the V3 API...

thecristen commented on apps/alerts/lib/cache/bus_stop_change_s3.ex:
```diff
@@ -0,0 +1,119 @@
+           Enum.filter(bus_alerts, &(&1.effect in [:stop_closure, :stop_moved])) do
+      Enum.map(stop_change_alerts, & &1.id)
+      |> Enum.sort()
+      |> cache(fn _sorted_alert_ids ->
```
The intent here was to avoid unnecessary rewrites of the same set of alert data.

thecristen replied:
Ah I think I see why it's a bit odd. I'm lowkey misappropriating the concache call all so I can skip redundant re-writes + avoid having to set up some alternate means of tracking the last rewrite myself. 😅

thecristen commented on apps/alerts/lib/cache/bus_stop_change_s3.ex:
```diff
@@ -0,0 +1,119 @@
+
+  @spec compress_alert(HistoricalAlert.t()) :: binary()
+  defp compress_alert(alert) do
+    :erlang.term_to_binary(alert, [:compressed])
```
Since I'm _only_ using these directly in a template with various functions that operate on an `%Alert{}`, it didn't feel worthwhile to add an extra JSON serializing/deserializing step I guess? :)

thecristen commented on apps/site/lib/site_web/controllers/bus_stop_change_controller.ex:
```diff
@@ -38,16 +38,34 @@ defmodule SiteWeb.BusStopChangeController do
     alerts = get_old_alerts()
+
     assign_bus_stop_alerts(conn, alerts)
+    |> assign(:past_alerts, past_stored_alerts())
```
It's not a _great_ reason, but boils down to weird interaction with the `SiteWeb.Plugs.AlertsByTimeframe` plug that I'm overly relying on to enable the past/current/upcoming filter buttons to the page.

thecristen commented on apps/alerts/lib/supervisor.ex:
```diff
@@ -24,7 +25,11 @@ defmodule Alerts.Supervisor do
+            {Alerts.Cache.Fetcher, api_mfa: @api_mfa},
+            BusStopChangeS3,
+            Supervisor.child_spec({Alerts.Cache.Fetcher, BusStopChangeS3.fetcher_opts()},
+              id: :bus_stop_change_worker
```
Good points - updated with hopefully something more appropriate.

paulswartz commented on apps/alerts/config/test.exs:
```diff
@@ -1,3 +1,5 @@
 
-config :alerts, api_mfa: {JsonApi, :empty, []}
+config :alerts,
+  api_mfa: {JsonApi, :empty, []},
```
question: is `api_mfa` used anywhere?

thecristen replied:
@paulswartz It's retrieved in `Alerts.Supervisor` to assign the module/function/args to `Alerts.Cache.Fetcher`.

paulswartz commented on apps/alerts/lib/historical_alert.ex:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  @spec get_name_or_id(String.t(), entity_key) :: String.t()
+  defp get_name_or_id(id, key) do
```
question: is this name right? I'm not sure it can return an ID.

thecristen replied:
Good catch, I think it survived a few refactors and should be fixed.

paulswartz commented on apps/alerts/test/historical_alert_test.exs:
```diff
@@ -0,0 +1,63 @@
+defmodule HistoricalAlertTest do
+  use ExUnit.Case, async: true
```
suggestion: as above, you shouldn't use `async: true` with mocks.
```suggestion
  use ExUnit.Case
```

paulswartz commented on apps/site/lib/site_web/views/bus_stop_change_view.ex:
```diff
@@ -110,7 +72,9 @@ defmodule SiteWeb.BusStopChangeView do
 
-  @spec time_range(%Alert{}) :: Phoenix.HTML.Safe.t()
+  @spec time_range(%Alert{} | %HistoricalAlert{}) :: Phoenix.HTML.Safe.t()
+  def time_range(%HistoricalAlert{alert: alert}), do: time_range(alert)
```
question: should this and `alert_item/1` below also be a part of the `AmbiguousAlert` interface?

paulswartz commented on apps/site/lib/site_web/ambiguous_alert.ex:
```diff
@@ -0,0 +1,63 @@
+defprotocol AmbiguousAlert do
```
nitpick: given the filesystem location, should this be `SiteWeb.AmbiguousAlert`?

paulswartz commented on apps/site/lib/site_web/views/bus_stop_change_view.ex:
```diff
@@ -1,80 +1,42 @@
     end)
-    |> Enum.uniq()
+    |> Enum.sort_by(&elem(&1, 0))
+    |> Enum.into(%{})
```
question: can you say more about what you're trying to do here? Maps aren't inherently ordered (and in fact the order is changing in Erlang 26), so if there's an order you're depending on this should be a List.

thecristen replied:
👍🏼   Yeah that doesn't work. Am trying to ensure everything is grouped _and_ sorted, as the template ultimately shows the alerts grouped by municipality, and grouped by stop name within each municipality. I've refactored this and everything's in Lists now.

thecristen commented on apps/alerts/test/alerts_test.exs:
thank you!!!

paulswartz commented on apps/site/lib/site_web/views/bus_stop_change_view.ex:
```diff
@@ -1,80 +1,70 @@
+    keyed_alerts
+    |> Enum.flat_map(&alert_with_related_stops/1)
+    |> sort_and_adjust_values(fn alerts_list ->
+      Enum.sort_by(alerts_list, &alert_start_date/1)
```
suggestion: when you're sorting by structs, you need to provide the name of the struct to ensure that they're not compared as maps.

```suggestion
      Enum.sort_by(alerts_list, &alert_start_date/1, DateTime)
```

There's more about why you need this in the `Enum.sort_by/3` [documentation](https://hexdocs.pm/elixir/Enum.html#sort_by/2).

paulswartz commented on apps/alerts/test/historical_alert_test.exs:
```diff
@@ -0,0 +1,63 @@
+defmodule HistoricalAlertTest do
```
nitpick:
```suggestion
defmodule Alerts.HistoricalAlertTest do
```

paulswartz reviewed (commented):
suggestion: I'm not sure if it's easier, but you might think about pulling out the `start_data_processes` commit(s) into a new PR.

thecristen commented on Dockerfile:
```diff
@@ -3,7 +3,7 @@
 
 # 1.) Get the Elixir dependencies within an Elixir container
-FROM hexpm/elixir:1.10.4-erlang-22.3.4.26-debian-buster-20210902 as elixir-builder
+FROM hexpm/elixir:1.12.3-erlang-22.3.4.26-debian-buster-20210902 as elixir-builder
```
whoops, ended up updating this in a different PR, will clean up here.

thecristen commented on apps/vehicles/lib/vehicles.ex:
```diff
@@ -12,9 +12,13 @@ defmodule Vehicles do
-      "USE_SERVER_SENT_EVENTS"
-      |> System.get_env()
-      |> stream_children()
+      if Application.get_env(:elixir, :start_data_processes) do
```
this was added in a prior PR that's since been merged, but will note this for a future tech debt task (sorry for the mess but thanks for the feedback!)

thecristen commented on apps/vehicles/lib/vehicles.ex:
```diff
@@ -12,9 +12,13 @@ defmodule Vehicles do
-      |> System.get_env()
-      |> stream_children()
+      if Application.get_env(:elixir, :start_data_processes) do
+        "USE_SERVER_SENT_EVENTS"
```
this was added in a prior PR that's since been merged, but will note this for a future tech debt task (sorry for the mess but thanks for the feedback!)

thecristen commented on apps/site/lib/site_web/views/bus_stop_change_view.ex:
```diff
@@ -1,80 +1,70 @@
+      {key, alerts_func.(alerts)}
     end)
+    # Sort whole list by keys
+    |> Enum.sort_by(&elem(&1, 0))
```
> The issue with the previous implementation wasn't that you used group_by, it was that you converted the sorted list back into a map which lost the ordering

Yep I got that, that's mainly why I stopped using `group_by` here, so maintaining it as a list throughout.
You're right though, I can deprecate `collect_by_group` with a `group_by`, will add that here :D

thecristen commented:
Thanks @paulswartz, I rebased this on the latest master (which had covered some of the ancillary stuff) and added one more commit to address the rest of your comments so far. ⚡

paulswartz reviewed (approved):
:cake:

Code looks good! Before merging, can you please deploy it to one of the dev environments to make sure there aren't any unexpected issues with deployment?

paulswartz commented on apps/alerts/lib/cache/bus_stop_change_s3.ex:
```diff
@@ -0,0 +1,128 @@
+  @ex_aws Application.compile_env(:alerts, [:mock_aws_client], ExAws)
+  @ex_aws_s3 Application.compile_env(:alerts, [:mock_aws_client], ExAws.S3)
+  @bucket "mbta-dotcom"
+  @bucket_prefix Application.compile_env!(:alerts, [:bus_stop_change_bucket])
```
@thecristen I'm surprised that this didn't generate a warning on startup, but you're getting the change prefix at compile time. That means you can't override it with the environment variable at runtime.

thecristen commented:
*finally* figured out the config; after adding the runtime config to the release _and_ replacing the module attribute with a function, it finally writes alerts to the bucket correctly!
